lib/lightning_web/live/credential_live/oauth_component.ex

```diff
@@ -0,0 +1,689 @@
+
+  require Logger
+
+  attr :form, :map, required: true
+  attr :id, :string, required: true
+  attr :update_body, :any, required: true
+  attr :provider, :any, required: true
+  slot :inner_block
+
+  def fieldset(assigns) do
+    changeset = assigns.form.source
+
+    parent_valid? = !(changeset.errors |> Keyword.drop([:body]) |> Enum.any?())
+
+    token_body_changeset =
+      Common.TokenBody.changeset(
+        changeset |> Ecto.Changeset.get_field(:body) || %{}
+      )
+
+    assigns =
+      assigns
+      |> assign(
+        update_body: assigns.update_body,
+        valid?: parent_valid? and token_body_changeset.valid?,
+        token_body_changeset: token_body_changeset
+      )
+
+    ~H"""
+    <%= render_slot(
+      @inner_block,
+      {Phoenix.LiveView.TagEngine.component(
+         &live_component/1,
+         [
+           module: __MODULE__,
+           form: @form,
+           token_body_changeset: @token_body_changeset,
+           update_body: @update_body,
+           provider: @provider,
+           id: "#{provider_name(@provider)}-oauth-inner-form-#{@id}"
+         ],
+         {__ENV__.module, __ENV__.function, __ENV__.file, __ENV__.line}
+       ), @valid?}
+    ) %>
+    """
+  end
+
+  @impl true
+  def render(%{client: nil} = assigns) do
+    ~H"""
+    <div id={@id} class="flex items-center place-content-center pt-2">
+      <div class="text-center">
+        <div class="text-base font-medium text-gray-900">
+          No Client Configured
+        </div>
+        <span class="text-sm">
+          <%= provider_name(@provider) |> String.capitalize() %> authorization has not been set up on this instance.
+        </span>
+      </div>
+    </div>
+    """
+  end
+
+  def render(assigns) do
+    assigns =
+      assigns
+      |> update(:form, fn form, %{token_body_changeset: token_body_changeset} ->
+        # Merge in any changes that have been made to the TokenBody changeset
+        # _inside_ this component.
+        %{
+          form
+          | params: Map.put(form.params, "body", token_body_changeset.params)
+        }
+      end)
+      |> assign(
+        show_authorize:
+          !(assigns.authorizing || assigns.error || assigns.userinfo)
+      )
+
+    ~H"""
+    <fieldset id={@id}>
+      <div :for={
+        body_form <- Phoenix.HTML.FormData.to_form(:credential, @form, :body, [])
+      }>
+        <%= Phoenix.HTML.Form.hidden_input(body_form, :access_token) %>
+        <%= Phoenix.HTML.Form.hidden_input(body_form, :refresh_token) %>
+        <%= Phoenix.HTML.Form.hidden_input(body_form, :expires_at) %>
+        <%= Phoenix.HTML.Form.hidden_input(body_form, :scope) %>
+        <%= Phoenix.HTML.Form.hidden_input(body_form, :instance_url) %>
+      </div>
+
+      <%= if @provider === Lightning.AuthProviders.Salesforce do %>
+        <.live_component
+          id={"#{@id}-scope-selection"}
+          parent_id={@id}
+          module={ScopeSelectionComponent}
+        />
+      <% end %>
```

Comment:
This generic oauth component shouldn't be aware of Salesforce, this should be injected in via a slot - or we go back to specific components for Google/Salesforce with a common module that imports (or a macro) that adds in the common behaviour.

Reply:
Moved it now as a function component and out of the `LightningWeb.CredentialLive.OauthComponent`.